kolibri/core/device/api.py

```diff
@@ -309,6 +314,37 @@ def get_queryset(self):
             return UserSyncStatus.objects.none()
         return UserSyncStatus.objects.all()
 
+    def downloads_queryset(self, queryset):
+
+        queryset = queryset.annotate(
+            last_download=Max("requested_at"), last_removal=Max("requested_at")
+        )
+
+        is_last_download_removed = Subquery(
+            ContentRemovalRequest.objects.filter(
+                source_id=OuterRef("user_id"),
+                source_model=FacilityUser.morango_model_name,
+                reason=ContentRequestReason.UserInitiated,
+            )
+        )
+
+        is_having_downloads = Subquery(
+            ContentDownloadRequest.objects.filter(
+                source_id=OuterRef("user_id"),
+                source_model=FacilityUser.morango_model_name,
+                reason=ContentRequestReason.UserInitiate,
+            )
+        )
+
+        queryset = queryset.annotate(
+            last_download_removed=Subquery(is_last_download_removed).values(
+                "last_removal"
+            ),
+            has_downloads=Subquery(is_having_downloads).values("last_download"),
+        )
+        pdb.set_trace()
```

Comment:
:eyes:

Reply:
As your test confirms, you've put forth some testing against this code. Although, I don't see `pdb` imported, leading me to suspect this code should raise an error if ever encountered. If it hasn't errored, then that must mean it isn't being invoked!

Reply:
This is actually a useful debugging practice during test development. If you want to sanity check an all-good, passing test result, you can throw an error, i.e. `raise new Exception('oops')`, which you expect to cause a clear failure.

Reply:
Looks like I just forgot to remove the `pdb` before pushing the code.